### PR TITLE
Resolve #20

### LINKED ISSUE #20: @CheckForNull and @Nonnull not working within lambdas
Original: https://github.com/findbugsproject/findbugs/issues/95 - reported by @lotsabackscatter

Null pointer de-reference checks do not appear to be working within Java 8 lambdas.

Anonymous classes work fine but converting these to lambdas effectively suppresses any warnings.

See below for an example test case:

```
package sfBugsNew;

import javax.annotation.CheckForNull;
import edu.umd.cs.findbugs.annotations.ExpectWarning;

/**
 */
public class Bug {

  /**
   * @return may be null
   */
  @CheckForNull
  public static String methodThatMightReturnNull() {
    if (Boolean.getBoolean("test")) {
      return null;
    }

    return "test";
  }

  /**
   */
  @ExpectWarning("NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE")
  public void testNPLambda() {
    Runnable lambda = () -> {
      String maybeNull = methodThatMightReturnNull();

      // would expect findbugs warning about possible null dereferencing here
      boolean empty = maybeNull.isEmpty();
      System.out.println(empty);
    };
    lambda.run();
  }

}
```

### PR DESCRIPTION
**Pull request by jsotuyod**
- Add the failing test scenario
 - Add a helper in BCELUtil to detect lambdas.
 - Avoid BCELUtil from returning true to `isSynthetic` for lambdas,
    not technically correct, but semantically, when detectors ask for
    `isSynthetic` they just want to know if the developer wrote the code
    and for lambdas, that is true.
 - Workaround the testing framework to deal with the lack of expectation
    warnings for lambdas. This needs a refactor.

Fixes #20

### CODE REVIEW

**Comment by jsotuyod** [comment]
Ok, this is probably not ready to be merged, but having the code available will certainly make discussion much easier.

I realize the change to `BCELUtility.isSynthetic` is bad. I am returning false if it's a lambda, even if it's a synthetic method. This is strictly broken, but semantically when asking for synthetic methods, what we want to ask when checking it is not "was this method created by the compiler?" but rather "was this method **not** written by the developer?". The difference is small, yet important. The body of a lambda method *was* written by the developer, and is therefore as interesting as any non-synthetic method.

What I wanted to discuss with you is the semantics we would like to offer for  proper fix to this.

We can't prevent plugin developers from accessing `isSynthetic` on any of the multiple classes and helpers that offer that method, but we may offer some higher-level abstractions such as `isAutogeneratedMethod` or `isLambda`; and refactor SpotBugs source to start using those instead of directly checking for synthetic fields / methods.

Not sure what to name such helper / where to put it, nor the semantics for those methods (is really `isAutogeneratedMethod` an appropriate name to tell apart accessors / autogenerated enum methods from lambdas and user methods?)

Also, this scenario shows once again the need of addressing #10, since the current testing integration is insufficient to properly / easily test these scenarios.

**Comment by jsotuyod** [comment]
Has anyone got the chance to think about this one?

**Comment by jsotuyod** [comment]
@henrik242 good call not merging this. It's marked as WIP, we still need to decide what are proper abstractions to present to detectors / plugin developers to make it easier to work with lambdas.

**Comment by henrik242** [comment]
@jsotuyod Yup. You should maybe change the title of this PR, though, since #20 is resolved :)

**Comment by jsotuyod** [comment]
Ok, a bunch of changes:

- I replaced the test to use the new test infrastructure (#10 & #35)
- I deprecated `BCELUtil.isSynthetic` (all versions), in favor of the corresponding higher level abstraction on the new `MemberUtils.isUserGenerated`. This change has already been impacted on `BugInstance`. Other detectors may be need to be updated, but should be analyzed separately.

**Comment by jsotuyod** [comment]
This is ready for review as far as I'm concerned. I feel the current abstraction is good enough.

File: findbugs/src/main/java/edu/umd/cs/findbugs/bytecode/MemberUtils.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+ * Utility to analyze class members.
+ */
+public final class MemberUtils {
+
+	private MemberUtils() {
+		throw new AssertionError("Utility classes can't be instantiated");
+	}
+
+	private static boolean internalIsSynthetic(final FieldOrMethod m) {
+        if (m.isSynthetic()) {
```
Comment (KengoTODA) [inline findbugs/src/main/java/edu/umd/cs/findbugs/bytecode/MemberUtils.java]: Checkstyle also reported problems:
```
[ant:checkstyle] [ERROR] /findbugs/src/main/java/edu/umd/cs/findbugs/SAXBugCollectionHandler.java:573:9: File contains tab characters (this is the first instance). [FileTabCharacter]
[ant:checkstyle] [ERROR] /findbugs/findbugs/src/main/java/edu/umd/cs/findbugs/Version.java:49:13: File contains tab characters (this is the first instance). [FileTabCharacter]
[ant:checkstyle] [ERROR] /findbugs/src/main/java/edu/umd/cs/findbugs/ba/generic/GenericUtilities.java:264:13: File contains tab characters (this is the first instance). [FileTabCharacter]
```

Comment (jsotuyod) [inline findbugs/src/main/java/edu/umd/cs/findbugs/bytecode/MemberUtils.java]: You are absolutely right. I'l fix this and have Checkstyle run on Travis too.

**Comment by jsotuyod** [comment]
There you go. I'll enable Checkstyle on a separate PR since there are a few other issues to solve and don't want to mix things on this PR.

**Pull request merged**